File: A_Question_Marks.py
def solve_test_case():
    n = int(input())
    s = input()
    
    count = {'A': 0, 'B': 0, 'C': 0, 'D': 0}
    question_marks = 0
    
    for c in s:
        if c == '?':
            question_marks += 1
        else:
            count[c] += 1
    
    if question_marks == 4*n:  # All questions are unanswered
        return 0
    
    correct = 0
    available = {'A': 0, 'B': 0, 'C': 0, 'D': 0}
    
    for letter in 'ABCD':
        correct += min(count[letter], n)
        if count[letter] < n:
            available[letter] = n - count[letter]
    
    additional_correct = 0
    for _ in range(question_marks):
        max_available = max(available.values())
        if max_available > 0:
            for letter in 'ABCD':
                if available[letter] == max_available:
                    available[letter] -= 1
                    additional_correct += 1
                    break
    
    return correct + additional_correct

t = int(input())
for _ in range(t):
    print(solve_test_case())

File: A_Question_Marks.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>

using namespace std;

int solve_test_case() {
    int n;
    string s;
    cin >> n >> s;
    
    vector<int> count(4, 0);
    int question_marks = 0;
    
    for (char c : s) {
        if (c == '?') {
            question_marks++;
        } else {
            count[c - 'A']++;
        }
    }
    
    int correct = 0;
    int remaining = 4 * n;
    
    for (int i = 0; i < 4; i++) {
        int correct_for_letter = min(count[i], n);
        correct += correct_for_letter;
        remaining -= count[i];
    }
    
    int additional_correct = min(question_marks, remaining);
    return correct + additional_correct;
}

int main() {
    int t;
    cin >> t;
    
    while (t--) {
        cout << solve_test_case() << endl;
    }
    
    return 0;
}